pkg/execution/pauses/block.go

```diff
@@ -0,0 +1,464 @@
+	if err != nil {
+		return fmt.Errorf("failed to generate block metadata: %w", err)
+	}
+
+	// Generate a deterministic ULID for the block ID based off of the last pause
+	// timestamp and ID in our block.
+	block.ID = blockID(block, metadata)
+
+	// Marshal the block.  We currently use JSON encoding everywhere, but
+	// we can amend Serialize to use protobuf if we desire via a new tag.
+	byt, err := Serialize(block, encodingJSON, 0x00)
+	if err != nil {
+		return fmt.Errorf("failed to serialize block: %w", err)
+	}
+
+	// Now that we have our block, write it to the backing store.
+	key := b.BlockKey(index, block.ID)
+	if err := b.bucket.WriteAll(ctx, key, byt, nil); err != nil {
+		return fmt.Errorf("failed to write block: %w", err)
+	}
+
+	// Write block index to our zset.
+	if err := b.addBlockIndex(ctx, index, block, metadata); err != nil {
+		return fmt.Errorf("failed to write block index: %w", err)
+	}
+
+	// Remove len(block.Pauses) from the buffer, as they've been flushed.
+	for _, p := range block.Pauses {
+		if err := b.buf.Delete(ctx, index, *p); err != nil {
+			logger.StdlibLogger(ctx).Warn("error deleting pause from buffer after flushing block", "error", err)
+		}
+	}
+
+	return nil
+}
+
+// BlocksSince returns all block IDs that have been written for a given index,
+// since a given time.
+//
+// If the time is ZeroTime, all blocks for the index must be returned.
+//
+// NOTE: This is NOT INCLUSIVE of since, ie. the range is (since, now].
+func (b blockstore) BlocksSince(ctx context.Context, index Index, since time.Time) ([]ulid.ULID, error) {
+	// Read from backing KV (redis/valkey/memorydb/fdb) indexes.
+	ms := since.UnixMilli()
+	score := "(" + strconv.Itoa(int(ms))
+	if since.IsZero() {
+		score = "-inf"
+	}
+
+	ids, err := b.rc.Do(
+		ctx,
+		b.rc.B().Zrangebyscore().Key(b.blockIndexKey(index)).Min(score).Max("+inf").Build(),
+	).AsStrSlice()
+	if err != nil {
+		return nil, fmt.Errorf("error querying block index: since %s: %w", score, err)
+	}
+
+	ulids := make([]ulid.ULID, len(ids))
+	for i, id := range ids {
+		ulids[i], err = ulid.Parse(id)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing block ULID '%s': %w", id, err)
+		}
+	}
+	return ulids, nil
+}
+
+func (b blockstore) ReadBlock(ctx context.Context, index Index, blockID ulid.ULID) (*Block, error) {
+	key := b.BlockKey(index, blockID)
+	byt, err := b.bucket.ReadAll(ctx, key)
+	if err != nil {
+		return nil, fmt.Errorf("error reading block from index '%s': id '%s': %w", index, blockID, err)
+	}
+
+	// Unmarshal the block, using the first byte to figure out encoding.
+	return Deserialize(byt)
+}
+
+// Delete deletes a pause from the backing blob.  Note that blobs are immutable;  we cannot
+// yeet the pause out of a blob as-is.  Instead, we track which blocks have deleted pauses
+// via indexes, and eventually compact blocks.
+func (b blockstore) Delete(ctx context.Context, index Index, pause state.Pause) error {
```

Comment:
should this operation be transactional?
what happens when there are network issues, io timeouts, context cancel, etc?
won't there be a ton of left over garbage?

also won't these much operations contribute to higher qps?

Reply:
On garbage: what garbage would you expect?  We're doing two reads, and inserting a pause into a single delete key - that single write is all we need to mark a deletion.

On transactions:  because this is a single write to delete, its not necessary.

Reply:
more so of a reaction when I see multiple interactions with redis in one function.
if this is fine, then we're good.